project-koku/koku
# RBAC implementation for AWS org units

PR for [COST-99](https://issues.redhat.com/browse/COST-99). Adding in RBAC for organizations api and report api for filtering/grouping by org units. 

Connected PR for the rbac config repo to update the cloud cost viewer role to have permissions: https://github.com/RedHatInsights/rbac-config/pull/62

###  Testing Instructions: 
Full access scenario: 
-  Inside of your `.env` make sure that you have the default org_admin development identity set: ```DEVELOPMENT_IDENTITY='{"identity": {"account_number": "10001", "type": "User", "user": {"username": "user_dev", "email": "user_dev@foo.com", "is_org_admin": "True", "access": {}}},"entitlements": {"cost_management": {"is_entitled": "True"}}}```
- Run `make load-aws-org-unit-tree`
- Test both the organizations API & the AWS costs endpoint: 
http://127.0.0.1:8000/api/cost-management/v1/organizations/aws/
http://127.0.0.1:8000/api/cost-management/v1/organizations/aws/?filter[org_unit_id]=OU_002
http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?group_by[org_unit_id]=R_001
http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?filter[org_unit_id]=OU_001
You should see 200s in the logs, as an example:
```
koku_server       | [2020-05-26 12:21:17,808] INFO None Identity: {"identity": {"account_number": "10001", "type": "User", "user": {"username": "user_dev", "email": "user_dev@foo.com", "is_org_admin": "True", "access": {}}}, "entitlements": {"cost_management": {"is_entitled": "True"}}}
koku_server       | [2020-05-26 12:21:17,808] INFO None {'method': 'GET', 'path': '/api/cost-management/v1/organizations/aws/?filter[org_unit_id]=R_001', 'request_id': None, 'account': '10001', 'username': 'user_dev', 'is_admin': 'True'}
koku_server       | [2020-05-26 12:21:17,884] WARNING None DEVELOPMENT is Enabled. Bypassing access lookup for user: {'identity': {'account_number': '10001', 'type': 'User', 'user': {'username': 'user_dev', 'email': 'user_dev@foo.com', 'is_org_admin': 'True', 'access': {}}}, 'entitlements': {'cost_management': {'is_entitled': 'True'}}}
koku_server       | [2020-05-26 12:21:18,046] INFO None {'method': 'GET', 'path': '/api/cost-management/v1/organizations/aws/?filter[org_unit_id]=R_001', 'status': 200, 'request_id': None, 'account': '10001', 'username': 'user_dev', 'is_admin': 'True'}
koku_server       | [2020-05-26 12:21:18,048] INFO None "GET /api/cost-management/v1/organizations/aws/?filter[org_unit_id]=R_001 HTTP/1.1" 200 7410
```


Limited & no access scenario 
-  Inside of your `.env` make sure that you have the following development identity set with access to the following orgs and accounts: ```DEVELOPMENT_IDENTITY='{"identity": {"account_number": "10001", "type": "User", "user": {"username": "user_dev", "email": "user_dev@foo.com", "access": {"aws.account": {"read": ["111111111111", "9999999999990", "9999999999990"]}, "aws.organizational_unit": {"read": ["R_001", "OU_001"]}}}}, "entitlements": {"cost_management": {"is_entitled": "True"}}}'```
- Test both the organizations API & the AWS costs endpoint: 
http://127.0.0.1:8000/api/cost-management/v1/organizations/aws/
http://127.0.0.1:8000/api/cost-management/v1/organizations/aws/?filter[org_unit_id]=OU_002
http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?group_by[org_unit_id]=R_001
http://127.0.0.1:8000/api/cost-management/v1/reports/aws/costs/?filter[org_unit_id]=OU_001
- When you hit the organizations endpoint without a filter - you should only see the `R_001` and `OU_001` organizations. When you filter on `OU_002` you should get a 403. Similarly both the filter and group_by above for the costs endpoint should return data - but if you change that to an org that you do not have permission to see, you will get a `403`, as an example: 
```
koku_server       | [2020-05-26 13:05:58,759] INFO None Identity: {"identity": {"account_number": "10001", "type": "User", "user": {"username": "user_dev", "email": "user_dev@foo.com", "access": {"aws.account": {"read": ["111111111111", "9999999999990", "9999999999990"]}, "aws.organizational_unit": {"read": ["R_001", "OU_001"]}}}}, "entitlements": {"cost_management": {"is_entitled": "True"}}}
koku_server       | [2020-05-26 13:05:58,759] INFO None {'method': 'GET', 'path': '/api/cost-management/v1/organizations/aws/?filter[org_unit_id]=OU_002', 'request_id': None, 'account': '10001', 'username': 'user_dev', 'is_admin': None}
koku_server       | [2020-05-26 13:05:58,813] WARNING None DEVELOPMENT is Enabled. Bypassing access lookup for user: {'identity': {'account_number': '10001', 'type': 'User', 'user': {'username': 'user_dev', 'email': 'user_dev@foo.com', 'access': {'aws.account': {'read': ['111111111111', '9999999999990', '9999999999990']}, 'aws.organizational_unit': {'read': ['R_001', 'OU_001']}}}}, 'entitlements': {'cost_management': {'is_entitled': 'True'}}}
koku_server       | [2020-05-26 13:05:58,839] WARNING None User does not have permissions for the requested params: {'OU_002'}. Current access: ['R_001', 'OU_001'].
koku_server       | [2020-05-26 13:05:58,864] INFO None {'method': 'GET', 'path': '/api/cost-management/v1/organizations/aws/?filter[org_unit_id]=OU_002', 'status': 403, 'request_id': None, 'account': '10001', 'username': 'user_dev', 'is_admin': None}
koku_server       | [2020-05-26 13:05:58,865] WARNING None Forbidden: /api/cost-management/v1/organizations/aws/
koku_server       | [2020-05-26 13:05:58,867] WARNING None "GET /api/cost-management/v1/organizations/aws/?filter[org_unit_id]=OU_002 HTTP/1.1" 403 6343
koku_server       | [2020-05-26 13:08:14,231] INFO None {'method': 'GET', 'path': '/api/cost-management/v1/organizations/aws/?filter[org_unit_id]=R_001', 'status': 200, 'request_id': None, 'account': '10001', 'username': 'user_dev', 'is_admin': None}
koku_server       | [2020-05-26 13:08:14,232] INFO None "GET /api/cost-management/v1/organizations/aws/?filter[org_unit_id]=R_001 HTTP/1.1" 200 7587
```
